community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/executionplan/PlanFingerprint.scala

```diff
@@ -23,23 +23,91 @@ import java.time.Clock
 
 import org.neo4j.cypher.internal.compiler.v3_2.spi.{GraphStatistics, GraphStatisticsSnapshot}
 
-case class PlanFingerprint(creationTimeMillis: Long, txId: Long, snapshot: GraphStatisticsSnapshot)
+case class PlanFingerprint(creationTimeMillis: Long, lastCheckTimeMillis: Long, txId: Long, snapshot: GraphStatisticsSnapshot)
 
-class PlanFingerprintReference(clock: Clock, minimalTimeToLive: Long, statsDivergenceThreshold : Double,
+class PlanFingerprintReference(clock: Clock, divergence: StatsDivergenceCalculator,
                                private var fingerprint: Option[PlanFingerprint]) {
 
-  def isStale(lastCommittedTxId: () => Long, statistics: GraphStatistics): Boolean = {
-    fingerprint.fold(false) { f =>
+  def isStale(lastCommittedTxId: () => Long, statistics: GraphStatistics): (Boolean, Int) = {
+    fingerprint.fold((false,0)) { f =>
       lazy val currentTimeMillis = clock.millis()
       lazy val currentTxId = lastCommittedTxId()
 
-      f.creationTimeMillis + minimalTimeToLive <= currentTimeMillis &&
-      check(currentTxId != f.txId,
-        () => { fingerprint = Some(f.copy(creationTimeMillis = currentTimeMillis)) }) &&
-      check(f.snapshot.diverges(f.snapshot.recompute(statistics), statsDivergenceThreshold),
-        () => { fingerprint = Some(f.copy(creationTimeMillis = currentTimeMillis, txId = currentTxId)) })
+      (divergence.shouldCheck(currentTimeMillis, f.lastCheckTimeMillis) &&
+        check(currentTxId != f.txId,
+          () => {
+            fingerprint = Some(f.copy(lastCheckTimeMillis = currentTimeMillis))
+          }) &&
+        check(f.snapshot.diverges(f.snapshot.recompute(statistics), divergence.decay(currentTimeMillis - f.creationTimeMillis)),
+          () => {
+            fingerprint = Some(f.copy(lastCheckTimeMillis = currentTimeMillis, txId = currentTxId))
+          }),
+        ((currentTimeMillis - f.lastCheckTimeMillis)/1000).toInt)
     }
   }
 
   private def check(test: => Boolean, ifFalse: () => Unit ) = if (test) { true } else { ifFalse() ; false }
 }
+
+trait StatsDivergenceCalculator {
+  val initialThreshold: Double
+  val initialMillis: Long
+
+  def shouldCheck(currentTimeMillis: Long, lastCheckTimeMillis: Long): Boolean = currentTimeMillis - initialMillis >= lastCheckTimeMillis
+
+  def decay(millisSincePreviousReplan: Long): Double
+}
+
+case class StatsDivergenceInverseDecayCalculator(initialThreshold: Double, targetThreshold: Double, initialMillis: Long, targetMillis: Long) extends StatsDivergenceCalculator {
```

Comment:
Do we really need two different calculators? How would we tell an end-user when to use one or the other?